types/ember__utils/-private/types.d.ts

```diff
@@ -2,6 +2,13 @@ export type KeysOfType<Base, Condition> = keyof Pick<Base, {
     [Key in keyof Base]: Base[Key] extends Condition ? Key : never
 }[keyof Base]>;
 
+// Since `TypeLookup` resolves all *other* types, including `null` and
+// `undefined`, we can assume that if the type does *not* resolve from
+// `KeysOfType`, it is safe to treat it as 'object'.
+export type TypeOf<Base, Condition> = KeysOfType<Base, Condition> extends never
```

Comment:
Note that this may have a small performance impact. The performance tests indicate it's small (as I would expect), but it's something to keep an eye on.

Reply:
This *should* be compatible with the extension in `@ember/object`, since that works by extending the `TypeLookup` interface.